lfs/gitscanner_catfilebatch.go

```diff
@@ -65,29 +60,73 @@ func runCatFileBatch(pointerCh chan *WrappedPointer, lockableCh chan string, loc
 	return nil
 }
 
-type catFileBatchScanner struct {
+type CatFileBatchScanner struct {
 	r       *bufio.Reader
-	blobSha string
-	pointer *WrappedPointer
-	err     error
+	w       io.Writer
+	closeFn func() error
+
+	blobSha     string
+	contentsSha string
+	pointer     *WrappedPointer
+	err         error
 }
 
-func (s *catFileBatchScanner) BlobSHA() string {
+func NewCatFileBatchScanner() (*CatFileBatchScanner, error) {
+	cmd, err := startCommand("git", "cat-file", "--batch")
+	if err != nil {
+		return nil, err
+	}
+
+	closeFn := func() error {
+		if err := cmd.Stdin.Close(); err != nil {
+			return err
+		}
+
+		stderr, _ := ioutil.ReadAll(cmd.Stderr)
+		if err := cmd.Wait(); err != nil {
+			return errors.Errorf("Error in git cat-file --batch: %v %v", err, string(stderr))
+		}
+
+		return nil
+	}
+
+	return &CatFileBatchScanner{
+		r:       cmd.Stdout,
+		w:       cmd.Stdin,
+		closeFn: closeFn,
+	}, nil
+}
+
+func (s *CatFileBatchScanner) BlobSHA() string {
 	return s.blobSha
 }
 
-func (s *catFileBatchScanner) Pointer() *WrappedPointer {
+func (s *CatFileBatchScanner) ContentsSha() string {
+	return s.contentsSha
+}
+
+func (s *CatFileBatchScanner) Pointer() *WrappedPointer {
 	return s.pointer
 }
 
-func (s *catFileBatchScanner) Err() error {
+func (s *CatFileBatchScanner) Err() error {
 	return s.err
 }
 
-func (s *catFileBatchScanner) Scan() bool {
+func (s *CatFileBatchScanner) Scan(sha []byte) bool {
```

Comment:
I like this improvement of calling `Scan(oid)`, instead of manually writing to stdin. Why does it take a `[]byte` though? The callers I saw get a sha as a string.